src/settings_gui.cpp

```diff
@@ -166,6 +167,178 @@ static const std::map<int, StringID> _volume_labels = {
+	NWidget(NWID_HORIZONTAL),
+		NWidget(WWT_FRAME, COLOUR_GREY, WID_GO_SOCIAL_PLUGIN_TITLE), SetDataTip(STR_JUST_STRING2, STR_NULL),
+			NWidget(NWID_HORIZONTAL), SetPIP(0, WidgetDimensions::unscaled.hsep_normal, 0),
+				NWidget(WWT_TEXT, COLOUR_GREY), SetMinimalSize(0, 12), SetFill(1, 0), SetDataTip(STR_GAME_OPTIONS_SOCIAL_PLUGIN_PLATFORM, STR_NULL),
+				NWidget(WWT_TEXT, COLOUR_GREY, WID_GO_SOCIAL_PLUGIN_PLATFORM), SetMinimalSize(100, 12), SetDataTip(STR_JUST_RAW_STRING, STR_NULL), SetAlignment(SA_RIGHT),
+			EndContainer(),
+			NWidget(NWID_HORIZONTAL), SetPIP(0, WidgetDimensions::unscaled.hsep_normal, 0),
+				NWidget(WWT_TEXT, COLOUR_GREY), SetMinimalSize(0, 12), SetFill(1, 0), SetDataTip(STR_GAME_OPTIONS_SOCIAL_PLUGIN_STATE, STR_NULL),
+				NWidget(WWT_TEXT, COLOUR_GREY, WID_GO_SOCIAL_PLUGIN_STATE), SetMinimalSize(100, 12), SetDataTip(STR_JUST_STRING1, STR_NULL), SetAlignment(SA_RIGHT),
+			EndContainer(),
+		EndContainer(),
+	EndContainer(),
+};
+
+static const NWidgetPart _nested_social_plugins_none_widgets[] = {
+	NWidget(NWID_HORIZONTAL),
+		NWidget(WWT_TEXT, COLOUR_GREY), SetMinimalSize(0, 12), SetFill(1, 0), SetDataTip(STR_GAME_OPTIONS_SOCIAL_PLUGINS_NONE, STR_NULL),
+	EndContainer(),
+};
+
+class NWidgetSocialPlugins : public NWidgetVertical {
+public:
+	NWidgetSocialPlugins()
+	{
+		this->plugins = SocialIntegration::GetPlugins();
+
+		if (this->plugins.size() == 0) {
+			auto widget = MakeNWidgets(std::begin(_nested_social_plugins_none_widgets), std::end(_nested_social_plugins_none_widgets), nullptr);
+			this->Add(std::move(widget));
+		} else {
+			for (size_t i = 0; i < this->plugins.size(); i++) {
+				auto widget = MakeNWidgets(std::begin(_nested_social_plugins_widgets), std::end(_nested_social_plugins_widgets), nullptr);
+				this->Add(std::move(widget));
+			}
+		}
+
+		this->SetPIP(0, WidgetDimensions::unscaled.vsep_wide, 0);
+	}
+
+	void FillWidgetLookup(WidgetLookup &widget_lookup) override
+	{
+		widget_lookup[WID_GO_SOCIAL_PLUGINS] = this;
+		NWidgetVertical::FillWidgetLookup(widget_lookup);
+	}
+
+	void SetupSmallestSize(Window *w) override
+	{
+		this->current_index = -1;
+		NWidgetVertical::SetupSmallestSize(w);
+	}
+
+	std::string &GetLongestSocialPlatform() const
+	{
+		std::string *longest = &this->plugins[0]->social_platform;
+		for (auto *plugin : this->plugins) {
+			if (plugin->social_platform.length() > longest->length()) longest = &plugin->social_platform;
+		}
+
+		return *longest;
+	}
+
+	std::string &GetLongestName() const
+	{
+		std::string *longest = &this->plugins[0]->name;
+		for (auto *plugin : this->plugins) {
+			if (plugin->name.length() > longest->length()) longest = &plugin->name;
+		}
+
+		return *longest;
+	}
+
+	std::string &GetLongestVersion() const
+	{
+		std::string *longest = &this->plugins[0]->version;
+		for (auto *plugin : this->plugins) {
+			if (plugin->version.length() > longest->length()) longest = &plugin->version;
+		}
+
+		return *longest;
+	}
```

Comment:
These need to use the longest pixel length, not the longest character length.

WWWWWWWW is wider than ..........

Reply:
It would really help me if you tell how, instead of just: this is wrong :P

Reply:
Normally I'd use `GetStringBoundingBox().width` in the window's `UpdateWidgetSize()` and return the width there, but pattern doesn't really work with your everything-in-a-custom-widget layout (which I really like btw).

So something like this, although maybe there's a std algorithm or some other pattern that can be used to reduce the duplication:
```
std::string *longest = &this->plugins[0]->version;
int longest_length = 0;
for (auto *plugin : this->plugins) {
    int length = GetStringBoundBox(plugin->version).width;
    if (length > longest_length) {
        longest_length = length;
        longest = &plugin->version;
    }
}

return *longest;
```

Reply:
Awesome, tnx! Templating power will take care of the rest :)